### PR TITLE
Lambda manager

### PR DESCRIPTION
**Pull request by jbeard4**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -160,3 +160,17 @@ pip-selfcheck.json
...
+python-orchestrator.zip
+python-worker.zip
+test-out.json
+test.json
+test13.json
+tests/aws_lambda/worker/commands.sh
+tests/aws_lambda/worker/function.zip
+tests/aws_lambda/worker/python-worker.zip
+tests/aws_lambda/worker/test.json
+tmp.py
```
Comment (paulmelnikow) [inline .gitignore:176]: Looks like a bunch of these aren't used anymore so most of it could be reset.

File: README.md [hunk 2]
```diff
@@ -193,6 +193,18 @@ You can monitor your queues using [RQ Dashboard][] or one of the
 [monitoring]: https://python-rq.org/docs/monitoring/
 
 
+### Parallel computation on AWS lambda
+
+Werkit also implements a parallel map on AWS lambda.
+
+Werkit comes with a default lambda handler, that accepts an event of the form `{"input":[a, b, ...],"extra_args":[c, d, ...]}`. Werkit invokes a lambda function in parallel for every item in `input`, with an event of the form `{"input": a, "extra_args":[c, d, ...]}`.
```
Comment (paulmelnikow) [inline README.md:200]: 👍

File: werkit/aws_lambda/default_handler.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import asyncio
+from .parallel import parallel_map_on_lambda
+from botocore.exceptions import ClientError
+
+LAMBDA_WORKER_FUNCTION_NAME = "LAMBDA_WORKER_FUNCTION_NAME"
+env_lambda_worker_function_name = (
+    os.environ[LAMBDA_WORKER_FUNCTION_NAME]
+    if LAMBDA_WORKER_FUNCTION_NAME in os.environ
+    else None
+)
```
Comment (paulmelnikow) [inline werkit/aws_lambda/default_handler.py:12]: Mappings have a convenience method for this:

```py
env_lambda_worker_function_name = os.environ.get("LAMBDA_WORKER_FUNCTION_NAME")
```

File: werkit/aws_lambda/default_handler.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+LAMBDA_WORKER_TIMEOUT = "LAMBDA_WORKER_TIMEOUT"
+env_lambda_worker_timeout = (
+    int(os.environ[LAMBDA_WORKER_TIMEOUT])
+    if LAMBDA_WORKER_TIMEOUT in os.environ
+    else None
+)
+
+
+def map_result(result):
```
Comment (paulmelnikow) [inline werkit/aws_lambda/default_handler.py:22]: I think `transform_result` would be a clearer name.

File: werkit/aws_lambda/default_handler.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+def handler(
+    event,
+    context,
+    lambda_worker_function_name=env_lambda_worker_function_name,
+    timeout=env_lambda_worker_timeout or 120,
+):
+    if not lambda_worker_function_name:
+        raise Exception(
+            f"Environment variable {LAMBDA_WORKER_FUNCTION_NAME} must be defined, or default kwArg lambda_worker_function_name must be bound to the handler"
```
Comment (paulmelnikow) [inline werkit/aws_lambda/default_handler.py:51]: 👍

File: werkit/aws_lambda/parallel.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        return ex
+    except asyncio.TimeoutError as ex:
+        return ex
+    except Exception as ex:
+        return ex
+
+
+async def parallel_map_on_lambda(
+    lambda_worker_function_name, timeout, input, extra_args=[]
+):
```
Comment (paulmelnikow) [inline werkit/aws_lambda/parallel.py:41]: Could you add a sentence of doc to these? At least to the one (I think it's this one, right?) that could be used without the orchestrator lambda.

File: werkit/aws_lambda/parallel.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+async def call_worker_service(lambda_worker_function_name, extra_args, _input):
+    async with aioboto3.client("lambda") as lambda_client:
+        # tic = time.time()
+        response = await lambda_client.invoke(
+            FunctionName=lambda_worker_function_name,
+            Payload=json.dumps({"input": _input, "extra_args": extra_args}),
+        )
+        # toc = time.time()
+        # print(_input, toc - tic)
```
Comment (paulmelnikow) [inline werkit/aws_lambda/parallel.py:20]: Could you remove?

File: werkit/aws_lambda/test_integration.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
+import os
+import base64
+import boto3
+import uuid
+import zipfile
+import tempfile
+import json
+from pprint import pprint
+
+role = "arn:aws:iam::139234625917:role/werkit-test-integration"
```
Comment (paulmelnikow) [inline werkit/aws_lambda/test_integration.py:10]: It seems like this rols probably should be pulled from an env var. For development we could use a `.env` file along with [python-dotenv](https://pypi.org/project/python-dotenv/).

File: werkit/aws_lambda/test_integration.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+path_to_orchestrator_zip = "/tmp/python-orchestrator.zip"
+path_to_worker_zip = "/tmp/python-worker.zip"
+
+
+# https://stackoverflow.com/a/1855118/366856
+def zipdir(path, ziph):
+    # ziph is zipfile handle
+    for root, dirs, files in os.walk(path):
+        for file in files:
+            ziph.write(os.path.join(root, file))
```
Comment (paulmelnikow) [inline werkit/aws_lambda/test_integration.py:22]: I think if you use [the `arcname` arg to `ziph.write()`](https://docs.python.org/3/library/zipfile.html#zipfile.ZipFile.write) you won't need to chdir before you call this.

e.g.
```py
             ziph.write(os.path.join(root, file), arcname=file)
```

File: werkit/aws_lambda/test_integration.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    worker_response = client.delete_function(FunctionName=worker_function_name)
+
+    orchestrator_response = client.delete_function(
+        FunctionName=orchestrator_function_name
+    )
+
+    return worker_response, orchestrator_response
+
+
+# tomorrow:
```
Comment (paulmelnikow) [inline werkit/aws_lambda/test_integration.py:108]: rm?

File: werkit/aws_lambda/test_parallel.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    default_timeout,
+)
+
+from asynctest import Mock, patch
+
+
+@patch("aioboto3.client", new_callable=Mock)
+def test_call_worker_service_success(mock_invoke):
+    # test that we expect to get a successful response out
+    # stubber = init()
```
Comment (paulmelnikow) [inline werkit/aws_lambda/test_parallel.py:32]: rm?

File: werkit/aws_lambda/test_util.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    m2.return_value = False
+    mock_invoke.return_value.__aenter__ = m
+    mock_invoke.return_value.__aexit__ = m2
+
+    m.return_value.invoke = CoroutineMock()
+
+    # setup mock
+    m.return_value.invoke.side_effect = invoke_return_values
+
+    return expected_result
```
Comment (paulmelnikow) [inline werkit/aws_lambda/test_util.py:44]: Wow, these mocks got a little complicated!

File: werkit/aws_lambda/test_worker/service.py [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import time
+import os
+
+LAMBDA_WORKER_DELAY = 'LAMBDA_WORKER_DELAY'
+env_lambda_worker_delay = int(os.environ[LAMBDA_WORKER_DELAY]) if LAMBDA_WORKER_DELAY in os.environ else 0
```
Comment (paulmelnikow) [inline werkit/aws_lambda/test_worker/service.py:5]:
```py
env_lambda_worker_delay = int(os.environ.get('LAMBDA_WORKER_DELAY', 0))
```

Could you include the units in the env var? e.g. `LAMBDA_WORKER_DELAY_SECONDS`?

File: .gitignore (hunk 1 above)
Comment (jbeard4) [reply .gitignore:176]: Fixed in 81d4c53dbbfbe047a2152da012c6607435c32954

File: werkit/aws_lambda/default_handler.py (hunk 3 above)
Comment (jbeard4) [reply werkit/aws_lambda/default_handler.py:12]: Fixed in 41247e71a1dac446ad6f22d12b88b2e23f5ced91

File: werkit/aws_lambda/default_handler.py (hunk 4 above)
Comment (jbeard4) [reply werkit/aws_lambda/default_handler.py:22]: Fixed in 40c1b1d2df915014df3bbdad532f34540f1c0a14

File: werkit/aws_lambda/test_integration.py (hunk 9 above)
Comment (jbeard4) [reply werkit/aws_lambda/test_integration.py:22]: Fixed in 5cbf2f38d0cebf4240abbc12753f29b27fc8e014

File: werkit/aws_lambda/test_parallel.py (hunk 11 above)
Comment (jbeard4) [reply werkit/aws_lambda/test_parallel.py:32]: Fixed in 5cbf2f38d0cebf4240abbc12753f29b27fc8e014

File: werkit/aws_lambda/test_worker/service.py (hunk 13 above)
Comment (jbeard4) [reply werkit/aws_lambda/test_worker/service.py:5]: Fixed in 5cbf2f38d0cebf4240abbc12753f29b27fc8e014

File: werkit/aws_lambda/test_integration.py (hunk 10 above)
Comment (jbeard4) [reply werkit/aws_lambda/test_integration.py:108]: Fixed in 5cbf2f38d0cebf4240abbc12753f29b27fc8e014

**Comment by jbeard4** [comment]
This PR was based on my fork of the repo, before I had write access. Please use this fork: https://github.com/metabolize/werkit/pull/44

**Pull request closed without merging**